### PR TITLE
[ALLUXIO-2120] Reset client contexts when master address changes during initialization.

### PR DESCRIPTION
**Pull request by uronce-cc**
https://alluxio.atlassian.net/browse/ALLUXIO-2120

Currently, when using hadoop compatible file system API, the client context is initialized only once. If during the first initialization, master address is invalid, then the invalid master address will be permanently used unless client JVM is restarted. A more user friendly behavior should be that if master address is reset to a different value, then the client context is reinitialized too.

### CODE REVIEW

**Comment by calvinjia** [comment]
@uronce-cc Thanks for the patch. Is it safe to reinitialize the context when a new authority is used, will the existing (possibly valid) connections be lost?

File: core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java [hunk 1]
```diff
@@ -449,6 +455,13 @@ public void initialize(URI uri, org.apache.hadoop.conf.Configuration conf) throw
     }
   }
 
+  private boolean noNeedToInitialize() {
+    InetSocketAddress currentMasterAddress = ClientContext.getMasterAddress();
+    boolean sameUri = mUri.getHost().equals(currentMasterAddress.getHostString()) &&
+          mUri.getPort() == currentMasterAddress.getPort();
```
Comment (uronce-cc) [inline core/client/src/main/java/alluxio/hadoop/AbstractFileSystem.java]: done

**Comment by uronce-cc** [comment]
jenkins, test this please

**Comment by uronce-cc** [comment]
@calvinjia The existing connections are closed when resetting the file system context, if there are RPCs executing, closing connections will block until all executing RPCs are done.

**Comment by uronce-cc** [comment]
For block level RPCs, like reading or writing a file, cached block store is now used in FileInStream and FileOutStream, so when new block store is reset in FileSystemContext, the previously created FileInStream and FileOutStream still uses the block store cached during creation.

**Comment by uronce-cc** [comment]
jenkins, test this please

**Comment by calvinjia** [comment]
@uronce-cc Thanks for the update, could you add unit tests to verify the behavior you mentioned (in-flight operations will not be affected by the change, and existing state will not be affected)? Also, who is responsible for cleaning up the existing connection pools which may be discarded after the reinitialization?

**Comment by uronce-cc** [comment]
I'll have another PR for trying a different approach to the problem. This PR tries to switch client contexts which is tricky.

**Pull request closed without merging**